src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1034,332 @@ begin
+  nondegenerate $ B.restrict $ B.orthogonal (submodule.span K ({x} : set V)) :=
+begin
+  refine λ m hm, submodule.coe_eq_zero.1 (hB₁ m.1 (λ n, _)),
+  have : n ∈ submodule.span K ({x} : set V) ⊔
+    B.orthogonal (submodule.span K ({x} : set V)) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw restrict_def at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+/- Let V be a finite dimensional vector space over the field K with the
+  nondegenerate bilinear form B. Then for all m ∈ M, f_m : M → R₁ : n ↦ B m n is
+  a linear functional in the dual space.
+
+  Furthermore, the map, φ : M → M* : m ↦ f_m is an isomorphism. -/
+
+/-- Given a bilinear form `B` on the modual `M`, `to_dual' B` is the linear map
+  from `M` to its dual such that `to_dual B m` is the functional `λ x, B m x`. -/
+def to_dual' (B : bilin_form R₃ M₃) : M₃ →ₗ[R₃] module.dual R₃ M₃ :=
+  B.to_lin
+
+@[simp] lemma to_dual'_def {B : bilin_form R₃ M₃} {m n : M₃} :
+  B.to_dual' m n = B m n := rfl
+
+lemma to_dual'_injective (B : bilin_form R₃ M₃) (hB : B.nondegenerate) :
+  function.injective (to_dual' B) :=
+B.to_dual'.to_add_monoid_hom.injective_iff.2 (λ a ha, hB _ (linear_map.congr_fun ha))
+
+open finite_dimensional
+
+variable [finite_dimensional K V]
+
+-- In order to show that `to_dual` is a surjective map we used the fact that
+-- the dimensions of a vector space equal to the dimensions of its dual.
+-- So rather than working with modules over rings, we work with vecotor spaces
+lemma to_dual'_bijective (B : bilin_form K V) (hB : B.nondegenerate) :
+  function.bijective (to_dual' B) :=
+begin
+  refine ⟨B.to_dual'_injective hB, _⟩,
+  change function.surjective B.to_dual',
+  refine (linear_map.injective_iff_surjective_of_findim_eq_findim
+    (linear_equiv.findim_eq _)).1 (B.to_dual'_injective hB),
+  have hB := classical.some_spec (exists_is_basis_finite K V),
+  haveI := classical.choice hB.2,
+  exact is_basis.to_dual_equiv _ hB.1
+end
+
+/-- To dual is the `linear_equiv` with the underlying linear map `to_dual'`. -/
+noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
+  V ≃ₗ[K] module.dual K V :=
+{ map_smul' := B.to_dual'.map_smul',
+  .. add_equiv.of_bijective B.to_dual'.to_add_monoid_hom (to_dual'_bijective B hB) }
+
+lemma to_dual_def {B : bilin_form K V} (hB : B.nondegenerate) {m n : V} :
+  B.to_dual hB m n = B m n := rfl
+
+-- We start proving that symmetric nondegenerat bilinear forms are diagonalisable, or equivilently
+-- there exists a orthogonal basis with respect to any symmetric nondegenerate bilinear form.
+
+lemma is_basis.trivial (hV : findim K V = 0) : is_basis K (λ x : fin 0, (0 : V)) :=
+begin
+  split,
+  { rw linear_independent_iff',
+    intros,
+    exact fin.elim0 i },
+  { rw ← findim_top at hV,
+    rw [eq_top_iff, (@findim_eq_zero K V _ _ _ _ _).1 hV],
+    exact bot_le }
+end
```

Comment:
Unless I'm missing something, this looks like it belongs in `finite_dimensional.lean`, as it doesn't say anything about bilinear forms. I'd be tempted to use `empty` instead of `fin 0`.

Reply:
I'm using `fin 0` since I'm inducting on `fin n` so the first case is `fin 0` and I don't think there is an easier to get `fin 0` from `empty`.  I will move the lemma to `finite_dimensional.lean` however